[SPMD] Refactor DSize to DSymInt to enable sym_numel

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99233
* #99232
* #99231
* __->__ #99206
* #99190

This commit uses `aten.arange.default` and `aten.arange.start` to
test `aten.sym_numel`.

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D45028715](https://our.internmc.facebook.com/intern/diff/D45028715)

mrshenli committed:
[SPMD] Refactor DSize to DSymInt to enable sym_numel

This commit uses `aten.arange.default` and `aten.arange.start` to
test `aten.sym_numel`.

[ghstack-poisoned]

yifuwang commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -59,13 +61,28 @@ class Schema:
-    size: int  # local size
-    placement: Placement  # DTensor placement of the dimension where the size was retrieved
+    value: int  # value that the SymInt evaluates to
+    op: torch._ops.OpOverload  # one of {sym_size, sym_numel}
```
nit: IIRC sym_size and sym_numel are `OverloadPackets` as opposed `OpOverload`.

yifuwang reviewed (approved):
LG!

mrshenli committed:
Update on "[SPMD] Refactor DSize to DSymInt to enable sym_numel"


This commit uses `aten.arange.default` and `aten.arange.start` to
test `aten.sym_numel`.

[ghstack-poisoned]

mrshenli reviewed (commented):
Test failure is irrelevant

```
test_decomp.py::TestDecompCUDA::test_comprehensive_pow_cuda_float32 Command took >30min, returning 124
FINISHED PRINTING LOG FILE of test_decomp 3/13 (/var/lib/jenkins/workspace/test/test-reports/test_decomp_ivs147t5.log)

test_decomp 3/13 failed!
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1389, in <module>
    main()
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1347, in main
    raise RuntimeError(
RuntimeError: test_decomp 3/13 failed!

Tip: You can keep running tests even on failure by passing --keep-going to run_test.py.
If running on CI, add the 'keep-going' label to your PR and rerun your jobs.

real	72m7.220s
user	96m18.810s
sys	23m19.876s
Error: Process completed with exit code 1.
```

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45028715).
